src/Poly.jl

```diff
@@ -2156,29 +2156,89 @@ end
 ###############################################################################
 
 @doc raw"""
-    evaluate(a::PolyRingElem, b::T) where T <: RingElement
+    evaluate(a::PolyRingElem, b::RingElement)
 
 Evaluate the polynomial expression $a$ at the value $b$ and return the result.
 """
-function evaluate(a::PolyRingElem, b::T) where T <: RingElement
+function evaluate(a::PolyRingElem, b::NCRingElement)
    i = length(a)
    R = base_ring(a)
    S = parent(b)
    if i == 0
       return zero(R) + zero(S)
    end
    if i > 25
-      return subst(a, b)
+      return evaluate_brent_kung(a, b)
+   end
+   return evaluate_horner(a, b)
+end
+
+function evaluate_horner(a::PolyRingElem, b)
+   i = length(a)
+   R = base_ring(a)
+   S = parent(b)
+   if i == 0
+      return zero(R) + zero(S)
    end
    z = coeff(a, i - 1) * one(S)
    while i > 1
       i -= 1
-      z = coeff(a, i - 1) + z*b
-      parent(z) # To work around a bug in julia
+      z = mul!(z, z, b)
+      z = add!(z, z, coeff(a, i - 1))
    end
    return z
 end
 
+# uses the algorithm of "A Practical Univariate Polynomial Composition Algorithm"
+# this is asymptotically faster, if scalar multiplications are cheaper than
+# base ring multiplications
+function evaluate_brent_kung(f::PolyRingElem{T}, a::U) where {T <: RingElement, U <: NCRingElement}
```

Comment:
This is also used when `a` is a (square) matrix, thus not a `NCRingElem`. So perhaps
```suggestion
function evaluate_brent_kung(f::PolyRingElem{T}, a::U) where {T <: RingElement, U}
```
?

Reply:
ah yes, thanks